go-controller/pkg/clustermanager/routeadvertisements/controller.go

```diff
@@ -280,6 +280,160 @@ func (c *Controller) reconcile(name string) error {
 	return c.updateRAStatus(ra, hadUpdates, err)
 }
 
+func subnetsOverlap(subnet1, subnet2 string) (bool, error) {
+	ipNets, err := util.ParseIPNets([]string{subnet1, subnet2})
+	if err != nil {
+		return false, err
+	}
+
+	// Two subnets overlap if one contains the other's IP or vice versa
+	return ipNets[0].Contains(ipNets[1].IP) || ipNets[1].Contains(ipNets[0].IP), nil
+}
+
+// checkSubnetOverlaps validates that subnets from the current RA don't overlap
+// with any existing subnets in the same VRFs, both within the RA itself and
+// across different RAs
+func (c *Controller) checkSubnetOverlaps(ra *ratypes.RouteAdvertisements, currentRASubnets []string) error {
+	// For now, we only check when not using "auto" mode (will handle "auto" later)
+	if ra.Spec.TargetVRF == "auto" {
+		return nil
+	}
+
+	targetVRF := ra.Spec.TargetVRF
+
+	// Check for overlaps within the current RA
+	if err := checkWithinRAOverlaps(ra.Name, targetVRF, currentRASubnets); err != nil {
+		return err
+	}
+
+	// Check for overlaps between current RA and other RAs in the same VRF
+	if err := c.checkCrossRAOverlaps(ra, targetVRF, currentRASubnets); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// checkWithinRAOverlaps checks if subnets within the current RA overlap with each other
+func checkWithinRAOverlaps(raName, targetVRF string, subnets []string) error {
+	for i, subnet1 := range subnets {
+		for j, subnet2 := range subnets {
+			if i >= j {
+				continue
+			}
+			overlap, err := subnetsOverlap(subnet1, subnet2)
+			if err != nil {
+				return fmt.Errorf("%w: failed to check subnet overlap between %s and %s: %w", errConfig, subnet1, subnet2, err)
+			}
+			if overlap {
+				return fmt.Errorf("%w: overlapping CIDR detected within RouteAdvertisements %q in VRF %q: %s and %s",
+					errConfig, raName, targetVRF, subnet1, subnet2)
+			}
+		}
+	}
+	return nil
+}
```

Comment:
You are assing raName and targetVRF just to print the error, that hsa to be done by the caller, 

create a helper multipleSubnetsOverlap with a []string as argument and use that at the caller instad of this function